jedi_language_server/server.py

```diff
@@ -724,6 +738,149 @@ def did_change_configuration(
+            MessageType.Debug,
+        )
+        return None
+
+    if len(definitions) > 1:
+        msg = (
+            f"multiple definitions found for name \"{n.description}\" of type '{n.type}' ({n.line}:{n.column}):\n"
+            f" {'\n'.join(map(lambda n: str(n), definitions))}"
+        )
+        server.show_message_log(msg, MessageType.Debug)
+
+    definition, *_ = definitions
+    if (
+        definition_type := SEMANTIC_TO_TOKEN_ID.get(definition.type, None)
+    ) is None:
+        server.show_message_log(
+            f"no matching semantic token for \"{n.description}\" of type '{n.type}' ({n.line}:{n.column})",
+            MessageType.Debug,
+        )
+        return None
+
+    return [n.line - 1, n.column, len(n.name), definition_type, 0]
+
+
+@SERVER.thread()
+@SERVER.feature(
+    TEXT_DOCUMENT_SEMANTIC_TOKENS_FULL,
+    SemanticTokensLegend(
+        token_types=SUPPORTED_SEMANTIC_TYPES, token_modifiers=[]
+    ),
+)
+def semantic_tokens_full(
+    server: JediLanguageServer, params: SemanticTokensParams
+) -> SemanticTokens:
+    """Thin wrap around  _semantic_tokens_range()."""
+    document = server.workspace.get_text_document(params.text_document.uri)
+    jedi_script = jedi_utils.script(server.project, document)
+
+    server.show_message_log(
+        f"semantic_tokens_full {params.text_document.uri} ",
+        MessageType.Log,
+    )
+
+    return _semantic_tokens_range(
+        server,
+        jedi_script,
+        Range(Position(0, 0), Position(INTEGER_MAX_VALUE, INTEGER_MAX_VALUE)),
+    )
+
+
+@SERVER.thread()
+@SERVER.feature(
+    TEXT_DOCUMENT_SEMANTIC_TOKENS_RANGE,
+    SemanticTokensLegend(
+        token_types=SUPPORTED_SEMANTIC_TYPES, token_modifiers=[]
+    ),
+)
+def semantic_tokens_range(
+    server: JediLanguageServer, params: SemanticTokensRangeParams
+) -> SemanticTokens:
+    """Thin wrap around  _semantic_tokens_range()."""
+    document = server.workspace.get_text_document(params.text_document.uri)
+    jedi_script = jedi_utils.script(server.project, document)
+
+    server.show_message_log(
+        f"semantic_tokens_range {params.text_document.uri} {params.range}",
+        MessageType.Log,
+    )
+
+    return _semantic_tokens_range(server, jedi_script, params.range)
+
+
+def _semantic_tokens_range(
+    server: JediLanguageServer, jedi_script: Script, doc_range: Range
+) -> SemanticTokens:
+    """General purpose function to do full / range semantic tokens."""
+    line, column = doc_range.start.line, doc_range.start.character
+    names = jedi_script.get_names(
+        all_scopes=True, definitions=True, references=True
+    )
+    data: list[int] = []
+
+    for n in names:
+        if (
+            not doc_range.start
+            < Position(n.line - 1, n.column)
+            < doc_range.end
+        ):
+            continue
+
+        token = _raw_semantic_token(server, n)
```

Comment:
For readability's sake, I think we should return a `NamedTuple` from `_raw_semantic_token`. We're doing enough manipulation of the returned value of `_raw_semantic_token` after this line where I think we can add the items to the end of `data` as a final step in the process.

As part of this, we can link to the LSP spec (either inline, or in this function's docstring) so that future readers quickly understand why we're doing such strange manipulations: 

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens